add icommentsmanger and icomment implementation

blizzz opened the pull request:
resolves #20409 
- [x] Register manager service in the server
- [x] offer config.php overwrite to replace default manager
- [x] listen for deleted users, run deleteReferencesOfActor()
- [x] listen for deleted files, run deleteCommentsAtObject()
- [x] ~~don't throw comments away immeditaly upon postDelete. The file may lay in trashbin and would be restored. Change object type to 'deleted_file' and look out for restore hook "post_restore" from "\OCA\Files_Trashbin\Trashbin". The file ID is stable.~~ Why bother?
- [x] ~~hook "delete" from "\OCP\Trashbin" indicates a file is permanently done. Only in this case delete the comments.~~ [unreliable](https://github.com/owncloud/core/pull/20775#issuecomment-161142958)
- [x] because of ^ remove listening to file hooks
- [x] fix Oracle lastInsertId() issue
- [x] Nice to have: consider runtime cache for comments in the manager

blizzz committed:
add icommentsmanger and icomment implementation

register CommentsManager service, allow override, document in config.sample.php

don't insert autoincrement ids in tests, because of dislikes from oracle and pgsql

specify timezone in null date

only accepts strings for ID parameter that can be converted to int

replace forgotten hardcoded IDs in tests

react on deleted users

react on file deletion

Postgresql compatibility

lastInsertId needs *PREFIX* with the table name

do not listen for file deletion, because it is not reliable (trashbin, external storages)

add runtime cache for comments

blizzz commented:
* [ ] Oracle

- Complains about inserting comments with specified id value. 
- Throws Doctrine\DBAL\Exception\DriverException when it does not like input that cannot be resolved to int
- does not like "0000-00-00"-initialized DateTimes, ``ORA-01841: (full) year must be between -4713 and +9999, and not be 0``

* [ ] Postgres
- Complains about inserting comments with specified id value with Doctrine\DBAL\Exception\InvalidFieldNameException
- Throws Doctrine\DBAL\Exception\DriverException when it does not like input that cannot be resolved to int
- does not like "0000-00-00"-initialized DateTimes, SQLSTATE[22007]: Invalid datetime format: 7 ERROR:  invalid input syntax for type timestamp: "-0001-11-30 00:00:00"

blizzz commented:
@DeepDiver1975 can we keep the test results longer, please? The Oracle and Postgre tests take ages until they start. Next time you have a look result details are already gone.

blizzz commented on lib/private/comments/manager.php:
```diff
@@ -0,0 +1,521 @@
+			/** @var \OCP\Files\Node $node */
+			if(isset($this->deleteList[$node->getPath()])) {
+				// hackish and possibly unreliable. What if a file is deleted without the preDelete hook?
+				$this->deleteCommentsAtObject('file', $this->deleteList[$node->getPath()]);
```
@icewind1991 is there a way I can find out the file id of a deleted file after it is actually deleted?

icewind1991 replied:
Not atm, but it shouldn't be to hard to add (for the node api hooks at least)

icewind1991 replied:
https://github.com/owncloud/core/pull/20865 allows doing a simple `$node->getId()` in the postDelete hook

blizzz replied:
K. Will this current approach work with external storages?

icewind1991 replied:
If the file is deleted from within ownCloud yes,

No approach will currently work if the file is removed on the external storage directly (same goes for shares, favorites, etc)

blizzz replied:
In the latter approach, will the entry from the file cache be removed at some point (e.g. when scanning)?

icewind1991 replied:
yes

blizzz replied:
talking to @schiesbn, '\OC\Files', 'postDelete' also happens when the file goes to trashbin. On restore, comments would be lost this way.

blizzz commented:
> hook "delete" from "\OCP\Trashbin" indicates a file is permanently done. Only in this case delete the comments.

Was thinking more about it. Since Trashbin is not a service per se, I may guess its emissions will be forwarded by the IRootFolder implementation (as usual delete hooks are)? However, since I don't know whether a trashbin is active (this implementation yes, but other storage wrappers…), I better really keep my fingers away from it and go for a background job that looks for orphaned file comments regularly and cleans up afterwards.

blizzz commented:
lastInsertId() is problematic in conjunction with Oracle. If I pass the tablename, it ends up with

```
14:11:05 1) Test_Comments_Manager::testGetComment
14:11:05 Doctrine\DBAL\Driver\OCI8\OCI8Exception: ORA-02289: sequence does not exist
14:11:05 
14:11:05 /ssd/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/OCI8/OCI8Exception.php:33
14:11:05 /ssd/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php:214
14:11:05 /ssd/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/OCI8/OCI8Connection.php:116
14:11:05 /ssd/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/OCI8/OCI8Connection.php:157
14:11:05 /ssd/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:1087
14:11:05 /ssd/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/lib/private/db/connection.php:226
14:11:05 /ssd/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/lib/private/db/adapteroci8.php:36
14:11:05 /ssd/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/lib/private/db/connection.php:221
14:11:05 /ssd/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/tests/lib/comments/manager.php:90
```

When I leave it out I am getting

```
Test_Comments_Manager::testGetComment
InvalidArgumentException: Oracle requires a table name to be passed into lastInsertId()
```

Did anybody ever came across this? @DeepDiver1975 @butonic maybe?

blizzz commented:
Review Time :grinning:  @nickvergessen @PVince81 @DeepDiver1975

PVince81 commented on lib/private/comments/comment.php:
```diff
@@ -0,0 +1,350 @@
+		'objectType'      => '',
+		'objectId'        => '',
+		'creationDT'      => null,
+		'latestChildDT'   => null,
```
Why not use actual PHP properties ?
I don't mind though, some of the FileInfo attributes are also stored that way.

blizzz replied:
makes life easier in fromArray()

PVince81 commented on lib/private/comments/manager.php:
```diff
@@ -0,0 +1,566 @@
+	) {
+		$this->dbConn = $dbConn;
+		$this->logger = $logger;
+		$userManager->listen('\OC\User', 'postDelete', function($user) {
```
The problem with this is that if no one used \OC::$server->getCommentsManager(), the hook will never be attached.

Maybe we can use the background job approach instead for easier/quicker orphans cleanup ? (like we do for regular tags)

blizzz replied:
There you have a point. Yes, we probably should put this into the cleanup background job as well.

nickvergessen replied:
No need for a background job, just register the hook in app.php instead, that is where it belongs to

blizzz replied:
this is not part of an app.  solved differently.

PVince81 commented on lib/private/comments/manager.php:
```diff
@@ -0,0 +1,566 @@
+	 * @throws \UnexpectedValueException
+	 */
+	protected function prepareCommentForDatabaseWrite(IComment $comment) {
+		if(    empty($comment->getActorType())
```
I think @Xenopathic said that `empty()` doesn't work when invoking functions and older PHP versions will complain, see https://github.com/owncloud/core/blob/v8.2.1/apps/files_external/service/globalstoragesservice.php#L106

PVince81 replied:
Maybe add a `validate()` or `isEmpty()` method on the $comment object instead ?

Xenopathic replied:
Yes, `empty()` takes a reference, so it will throw warnings if passed a transitive value like from a function return call. However, as I've pointed out in numerous places, `if(empty(...))` === `if(!...)`, as in, they are literally identical.So there is no need for `empty()` at all.

blizzz replied:
Oh, true, I forgot this… run into the same trap once upon a time :fearful: Thanks for finding this.

PVince81 commented on lib/private/comments/manager.php:
```diff
@@ -0,0 +1,566 @@
+		while($data = $resultStatement->fetch()) {
+			$comment = new Comment($this->normalizeDatabaseData($data));
+			$this->cache($comment);
+			$tree['replies'][] = [
```
I guess we don't cache replies?

blizzz replied:
We'd need to cache the sole tree. Could work, as long as we invalidate the cache when a child is added or deleted, depending on it's (topmost)ParentId. However,  getTree() could go against some "subParent", theoretically, in which case we cannot really invalidate the cache, or empty the whole thing. With offset and limit paremters, this is getting more difficult. 

So, for now,  we don't.

PVince81 replied:
Sounds good

PVince81 commented on lib/private/comments/manager.php:
```diff
@@ -0,0 +1,566 @@
+		    || !is_string($objectId)   || empty($objectId)
+		) {
+			// unsure whether it's a good place to enforce it here, since the
+			// comment instance can be manipulated anyway.
```
Maybe enforce on the $comment class directly using `validate()`

blizzz replied:
The Comment would already throw Exceptions on invalid parameters, although it does not check for emptiness. This would mean rely on the  IComment implementation, but since it's ours and cannot be replaced other than reimplementing the manager we can indeed do that. I chose the extremely defensive approach here, but  it is superfluous, I agree.

PVince81 commented on lib/private/comments/manager.php:
```diff
@@ -0,0 +1,566 @@
+	 */
+	public function save(IComment &$comment) {
+		$comment = $this->prepareCommentForDatabaseWrite($comment);
+		if($comment->getId() === '') {
```
why not "null" or empty ?

blizzz replied:
Id must be a string, this is enforced. Empty() can be used as well but won't make a difference.

PVince81 commented on lib/private/comments/manager.php:
```diff
@@ -0,0 +1,566 @@
+		$affectedRows = $qb
+			->update('comments')
+			->set('actor_type',	$qb->createNamedParameter(ICommentsManager::DELETED_USER))
+			->set('actor_id',	$qb->createNamedParameter(ICommentsManager::DELETED_USER))
```
I originally thought we'd always have a user id here and then call `UserManager->getUser()` on it to get the user's display name ? And if the user doesn't exist any more, show the deleted user text instead ?

Might be expensive though. @nickvergessen does the activity app resolve user display names ?

nickvergessen replied:
Only live, so for old users the ID is shown

nickvergessen replied:
But I'd agree, we should retain the name instead of killing the info, so in case a user is restored, it is at least possible to fix that again

blizzz replied:
A non existing user does not need to be deleted, but might be just temporarily unavailable (e.g. LDAP down for maintenance). We agreed that for deleted users "Deleted user" should be displayed. However we did not discuss how to deal with not present users. 

It's probably simpler to go with this approach and just display "Deleted user" if the user manager responds with null to a get().  Then we don't need the whole method and can kick out of the interface just as well.

blizzz replied:
> But I'd agree, we should retain the name instead of killing the info, so in case a user is restored, it is at least possible to fix that again

That's totally how we not do it in  ownCloud. If a user gets deleted, everything is removed,  files, preferences, etc. This also avoids that a new user with  the same username that accidentally another one had, suddenly has data that do not belong to him.

nickvergessen replied:
well that is why you change the type, so you know its no actual user anymore

blizzz replied:
Still I don't need to keep the ID either. We don't have user restore mechanism.

PVince81 replied:
So does it mean that if a user's display name changes, it won't change in the older comments ?

blizzz replied:
why user display name? That's a dynamic value that can change any time and is totally up to the discretion of the user backend. The only thing is stored is the userid, which is fixed and cannot change. The whatever client requests the display name from the user management by passing the user id. We don't have to do anything with this.

PVince81 replied:
Never mind, I misunderstood.

This is already storing the user id. At display time we'll need to resolve it to the proper display name anyway.

blizzz replied:
If a user get's deleted both type and id change to "deleted_user". The client's task is to show the appropriate display name. For the 'user'-type its to retrieve the display name, for 'deleted_user' types it is to show "Deleted user" instead (cf. https://github.com/owncloud/core/issues/16328#issuecomment-157312323).

PVince81 replied:
Fair enough :smile:

PVince81 commented on lib/private/comments/managerfactory.php:
```diff
@@ -0,0 +1,24 @@
+	 */
+	public function getManager() {
+		return new Manager(
+			\oc::$server->getDatabaseConnection(),
```
Big `OC` ? :wink:

blizzz replied:
hm…

PVince81 commented on lib/public/comments/icommentsmanagerfactory.php:
```diff
@@ -0,0 +1,22 @@
+/**
+ * Interface IComment
+ *
+ * This class represents a comment and offers methods for modification.
```
update this phpdoc

PVince81 replied:
@blizzz forgot this one ? :wink:

blizzz replied:
Oh, kinda, I improved one, but in the wrong file :dancer:

PVince81 commented on tests/lib/comments/fakefactory.php:
```diff
@@ -0,0 +1,22 @@
+ * For mock creation and auto-loading it extends Test\TestCase. I am, uh, really
+ * sorry for this hack.
+ */
+class Test_Comments_FakeFactory extends Test\TestCase implements \OCP\Comments\ICommentsManagerFactory {
```
Does it have to extend "TestCase" ? I'd suggest to remove it and then also remove `testNothing`

blizzz replied:
see the comment above

PVince81 commented on tests/lib/comments/fakefactory.php:
```diff
@@ -0,0 +1,22 @@
+	}
+
+	public function getManager() {
+		return $this->getMock('\OCP\Comments\ICommentsManager');
```
You should actually be able to remove this class completely and write a mock for `ICommentsManagerFactory` and make that mock return the other mock for `ICommentsManager`

blizzz replied:
Tried that first, but it did not work. IIRC because the class name could not be resolved.

PVince81 replied:
It should work, maybe you mixed up the mock calls ?

PVince81 commented on tests/lib/comments/manager.php:
```diff
@@ -0,0 +1,560 @@
+		\oc::$server->getDatabaseConnection()->prepare($sql)->execute();
+	}
+
+	protected function addDatabaseEntry($parentId, $topmostParentId, $creationDT = null, $latestChildDT = null) {
```
would it be possible to use the real actual method ?

Because this is quite a lot of code here and it would have to be kept in sync with the actual implementation in case of changes. And you'd need a test for the test, and then a test for the test of the test :smile:

blizzz replied:
Relying on the class to be tested while testing it? If the original method is bogus and renders all test cases to work, although they don't?

PVince81 replied:
That's a good point, but I think it won't happen because if you add enough tests, if adding comments is broken we will notice when trying to read them. If reading comments is broken, it will be noticed in the test that adds them.

PVince81 replied:
(I did this for the tags tests)

nickvergessen replied:
well first test is testing "add", empty table, call it once, check table, there should be one result.
and then you can use that method?

PVince81 commented on tests/lib/comments/manager.php:
```diff
@@ -0,0 +1,560 @@
+		$latestChildDT = new \DateTime('yesterday');
+
+		$qb = \oc::$server->getDatabaseConnection()->getQueryBuilder();
+		$qb
```
use the actual implementation instead.

blizzz replied:
what if the original methods would just cache it instead of writing to the DB? This would not be detected in testSaveNew() because i use the implementations there.

PVince81 replied:
True.

Anyway, it was just a suggestion. Feel free to keep this approach if it works for you.

MorrisJobke commented:
@blizzz I changed the state from To Review to Developing - see @PVince81's comments ;)

nickvergessen commented on tests/lib/comments/manager.php:
```diff
@@ -0,0 +1,560 @@
+ *
+ * @group DB
+ */
+class Test_Comments_Manager extends Test\TestCase
```
please use `namespace Test\Comments;`
For better file name convenience let the test file name end with test.php

nickvergessen commented on tests/lib/comments/comment.php:
```diff
@@ -0,0 +1,102 @@
+<?php
+
+class Test_Comments_Comment extends Test\TestCase
```
please use `namespace Test\Comments;`
For better file name convenience let the test file name end with test.php

blizzz committed:
different strategy in cleaning up after user was deleted

we do not listen to deletion hooks anymore, because there is no guarantee that they
will be heard - requires that something fetches the CommentsManager first.

Instead, in the user deletion routine the clean up method will be called directly. Same way
as it happens for files, group memberships, config values.

blizzz committed:
fix usage of empty

blizzz committed:
parameter checks for setting actor and object to happen only in one place

blizzz committed:
OC not oc

blizzz committed:
fix php doc

blizzz committed:
namespaces for tests

blizzz commented:
All open tasks addressed. I squashed the earlier commits into  one, now only changes to address the points made are kept in single commits.

The biggest change is probably about user deletion. I removed the listener, but added the call in the user deletion routine, where many others are, cf.  https://github.com/owncloud/core/pull/20775#discussion_r46565803

blizzz committed:
rework test about overwriting default comments manager

blizzz commented:
i rewrote the unsatisfying test on overwriting the defaults comments manager.

PVince81 commented on lib/private/comments/manager.php:
```diff
@@ -0,0 +1,550 @@
+			|| !$comment->getActorId()
+			|| !$comment->getObjectType()
+			|| !$comment->getObjectId()
+			|| !$comment->getVerb()
```
add a method `isEmpty()` on the comment class ? Then it can be used in other places as well.

blizzz replied:
I don't see the necessity for another method. The methods all returns strings, no surprises. The only thing i might should add is trim() in the setters.

PVince81 commented:
Looks good :+1:

PVince81 commented:
@DeepDiver1975 @nickvergessen second review ?

blizzz committed:
fix phpdoc

blizzz committed:
hardening, add some checks for whitespace-only strings

blizzz commented:
failing pgsql test is unrelated

blizzz commented:
Would be nice to get this merged today before you all switch to the maintenance release stuff next week :pray:

PVince81 commented:
@icewind1991 mind reviewing this?

icewind1991 commented on lib/private/comments/manager.php:
```diff
@@ -0,0 +1,550 @@
+		$data['id'] = strval($data['id']);
+		$data['parent_id'] = strval($data['parent_id']);
+		$data['topmost_parent_id'] = strval($data['topmost_parent_id']);
+		$data['creation_timestamp'] = new \DateTime($data['creation_timestamp']);
```
iirc you need to use `'@' . $timestamp` or `DateTime::createFromFormat('U', $timestamp)`

blizzz replied:
um, why? Tests are passing.

icewind1991 commented on lib/private/comments/manager.php:
```diff
@@ -0,0 +1,550 @@
+	 */
+	public function get($id) {
+		if(intval($id) === 0) {
+			throw new \InvalidArgumentException('IDs must be translatable to a number in this implementation.');
```
cant we just throw a notfound exception if it's not numeric since we only generate numeric ids, meaning that a non numeric id can't exist

blizzz replied:
In theory, yes. In practice, this offers more information if the case should occur.

icewind1991 commented on lib/private/comments/manager.php:
```diff
@@ -0,0 +1,550 @@
+	 * @return IComment
+	 * @throws \UnexpectedValueException
+	 */
+	protected function prepareCommentForDatabaseWrite(IComment $comment) {
```
Note that you don't actually return a new comment object here which might be confusing since I would expect that looking at the function signature

blizzz replied:
OK, I'll adjust the PHP doc

blizzz replied:
addressed

icewind1991 commented on lib/private/comments/manager.php:
```diff
@@ -0,0 +1,550 @@
+	 * @throws NotFoundException
+	 * @since 9.0.0
+	 */
+	public function save(IComment &$comment) {
```
Since (the current implementation of) `prepareCommentForDatabaseWrite` doesn't actually return a *new* comment object you don't need to pass `$comment` by reference since this function currently mutates the passed argument

blizzz replied:
The by-reference indicator is put into the interface deliberately as indicator, because the provided instance is supposed to receive the ID when saving is successful. Since we're passing around instances anyway, it's technically superfluous. Apparently it's also more confusing. I gonna remove it as well.

blizzz replied:
addressed

icewind1991 commented on lib/private/server.php:
```diff
@@ -1122,6 +1129,13 @@ public function getNotificationManager() {
 	/**
+	 * @return \OCP\Comments\ICommentsManager
+	 */
+	public function getCommentsManager() {
```
Needs to be exposed in `IServerContainer`?

blizzz replied:
good catch, thx

blizzz committed:
anounce CommentsManager getter in public server interface

blizzz committed:
improve PHP doc and remove superflous by reference indicator

blizzz commented:
*rebased*

blizzz committed:
adjust test's fakemanager to interface change

blizzz commented:
failing tests in core-ci-linux-swift-primary-storage/database=mysql,label=SLAVE unrelated

icewind1991 commented:
:+1: lets get this in

DeepDiver1975 commented on tests/lib/comments/manager.php:
```diff
@@ -0,0 +1,550 @@
+
+	public function testGetCommentNotFoundInvalidInput() {
+		$manager = $this->getManager();
+		$this->setExpectedException('\InvalidArgumentException');
```
use annotation please

blizzz replied:
what's the difference?

DeepDiver1975 replied:
I prefer to have it as annotation since we use it everywhere

DeepDiver1975 commented:
break on php 5.5

![bildschirmfoto von 2015-12-09 16-08-36](https://cloud.githubusercontent.com/assets/1005065/11688834/2bc39962-9e8f-11e5-8c26-47f441a9886d.png)

DeepDiver1975 commented:
![bildschirmfoto von 2015-12-09 16-09-32](https://cloud.githubusercontent.com/assets/1005065/11688858/423b5d7e-9e8f-11e5-9809-fed02c4209e4.png)
![bildschirmfoto von 2015-12-09 16-09-22](https://cloud.githubusercontent.com/assets/1005065/11688859/423c31cc-9e8f-11e5-8af6-686af7ac9e10.png)

DeepDiver1975 commented on lib/private/comments/manager.php:
```diff
@@ -0,0 +1,550 @@
+namespace OC\Comments;
+
+use Doctrine\DBAL\Exception\DriverException;
+use OC\Hooks\Emitter;
```
unused

DeepDiver1975 commented on tests/lib/comments/fakemanager.php:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Class FakeManager
+ */
+class FakeManager implements \OCP\Comments\ICommentsManager {
```
can't we use mocks for this?

blizzz replied:
no. see discussion.

blizzz commented:
m(

DeepDiver1975 commented:
@blizzz pro tip - tell PHPStorm to check for certain php version

![bildschirmfoto von 2015-12-09 16-17-57](https://cloud.githubusercontent.com/assets/1005065/11689084/7e97c2fc-9e90-11e5-9226-bb4f5be2f539.png)

nickvergessen commented on lib/private/comments/manager.php:
```diff
@@ -0,0 +1,550 @@
+			->execute();
+
+		if ($affectedRows > 0) {
+			$comment->setId(strval($this->dbConn->lastInsertId('*PREFIX*comments')));
```
`$qb->getLastInsertId()` at your service as of today, then you don't need to do casting of the value and no prefix

blizzz replied:
it returns an int so i need to convert it to string, but otherwise it's nicer and short. Thanks for the hint.

nickvergessen commented on tests/lib/comments/manager.php:
```diff
@@ -0,0 +1,550 @@
+			])
+			->execute();
+
+		return \OC::$server->getDatabaseConnection()->lastInsertId('*PREFIX*comments');
```
`$qb->getLastInsertId()` at your service as of today, then you don't need to do casting of the value and no prefix

nickvergessen commented on tests/lib/comments/manager.php:
```diff
@@ -0,0 +1,550 @@
+			])
+			->execute();
+
+		$id = strval(\OC::$server->getDatabaseConnection()->lastInsertId('*PREFIX*comments'));
```
`$qb->getLastInsertId()` at your service as of today, then you don't need to do casting of the value and no prefix

blizzz committed:
php < 5.5 compatible

blizzz committed:
remove unused use statement

blizzz committed:
use getLastInsertId from query builder for convenience

blizzz committed:
use expectedException annotation, not via method call, for consistency

blizzz commented:
@DeepDiver1975 your remarks are addressed

DeepDiver1975 commented:
:+1:
